crates/sound/src/audio_backends/pipewire.rs

```diff
@@ -1,30 +1,566 @@
+                                stream2.buffers.pop_front();
+                            }
+                        }
+
+                        (*(*datas).chunk).offset = 0;
+                        (*(*datas).chunk).stride = frame_size as i32;
+                        (*(*datas).chunk).size = n_bytes;
+
+                        stream.queue_raw_buffer(b);
+                    }
+                })
+                .register()
+                .expect("failed to register stream listener");
+
+            stream_listener.insert(stream_id, listener_stream);
+
+            let direction = match stream_params[stream_id as usize].direction {
+                VIRTIO_SND_D_OUTPUT => spa::Direction::Output,
+                VIRTIO_SND_D_INPUT => spa::Direction::Input,
+                _ => panic!("Invalid direction"),
+            };
+
+            stream
+                .connect(
+                    direction,
+                    Some(pipewire::constants::ID_ANY),
+                    pw::stream::StreamFlags::RT_PROCESS
+                        | pw::stream::StreamFlags::AUTOCONNECT
+                        | pw::stream::StreamFlags::INACTIVE
+                        | pw::stream::StreamFlags::MAP_BUFFERS,
+                    &mut param,
+                )
+                .expect("could not connect to the stream");
+
+            self.thread_loop.unlock();
+
+            // insert created stream in a hash table
+            stream_hash.insert(stream_id, stream);
+        }
+
+        Ok(())
+    }
+
+    fn release(&self, stream_id: u32, mut msg: ControlMessage) -> Result<()> {
+        debug!("pipewire backend, release function");
+        let release_result = self.stream_params.write().unwrap()[stream_id as usize]
+            .state
+            .release();
+        if let Err(err) = release_result {
+            log::error!("Stream {} release {}", stream_id, err);
+            msg.code = VIRTIO_SND_S_BAD_MSG;
+        } else {
+            self.thread_loop.lock();
+            let mut stream_hash = self.stream_hash.write().unwrap();
+            let mut stream_listener = self.stream_listener.write().unwrap();
+            let st_buffer = &mut self.stream_params.write().unwrap();
+
+            if let Some(stream) = stream_hash.get(&stream_id) {
+                stream.disconnect().expect("could not disconnect stream");
+                std::mem::take(&mut st_buffer[stream_id as usize].buffers);
+            } else {
+                return Err(Error::StreamWithIdNotFound(stream_id));
+            };
+            stream_hash.remove(&stream_id);
+            stream_listener.remove(&stream_id);
+
+            self.thread_loop.unlock();
+        }
+
+        Ok(())
+    }
+
+    fn start(&self, stream_id: u32) -> Result<()> {
+        debug!("pipewire start");
+        let start_result = self.stream_params.write().unwrap()[stream_id as usize]
+            .state
+            .start();
+        if let Err(err) = start_result {
```

Comment:
this looks wrong, as you are not propagating the error unless that is intended

Reply:
it's intended, we don't want it to panic in that case, logging the error should be enough.

Reply:
Aha, that is the second time i comment about the same thing. Definitely worth a comment then :)